Gems/ROS2/Code/Source/Manipulator/JointPublisherComponent.cpp

```diff
@@ -0,0 +1,126 @@
+#include <ROS2/Manipulator/JointPublisherComponent.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <ROS2/ROS2Bus.h>
+#include <PhysX/Joint/PhysXJointRequestsBus.h>
+#include <Source/HingeJointComponent.h>
+#include <AzCore/Component/ComponentApplicationBus.h>
+#include <ROS2/Frame/ROS2FrameComponent.h>
+#include <RobotImporter/URDFMetadataComponent.h>
+#include <ROS2/Utilities/ROS2Names.h>
+
+namespace ROS2
+{
+    void JointPublisherComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+        auto ros2Node = ROS2::ROS2Interface::Get()->GetNode();
+        auto ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
+        AZStd::string namespacedTopic = ROS2Names::GetNamespacedName(ros2Frame->GetNamespace(), "joint_states");
+        m_jointstatePublisher = ros2Node->create_publisher<sensor_msgs::msg::JointState>(namespacedTopic.data(), rclcpp::QoS(1));        // TODO: add QoS instead of "1"
+    }
+
+    void JointPublisherComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+        m_jointstatePublisher.reset();
+    }
+
+    void JointPublisherComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        provided.push_back(AZ_CRC_CE("JointPublisher"));
+    }
+
+    void JointPublisherComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("URDFMetadata"));
+    }
+
+    void JointPublisherComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<JointPublisherComponent, AZ::Component>()
+                ->Version(0);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<JointPublisherComponent>("JointPublisherComponent", "[Publish all the Hinge joint in the tree]")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2");
+            }
+        }
+    }
+
+    void JointPublisherComponent::InitializeMap()
+    {
+        if (auto* metadataComponent = GetEntity()->FindComponent<URDFMetadataComponent>())
+        {
+            m_hierarchyMap = metadataComponent->GetHierarchy();
+            AZ_TracePrintf("JointPublisherComponent", "Map initialized");
+        }
```

Comment:
What will happen if we can't initialze (find) the map? Maybe add an AZ_Assert here as well